libstore: generalize signature scheme and introduce remote signing

RaitoBezarius opened the pull request:
# Motivation

As discussed in https://github.com/NixOS/nixos-org-configurations/issues/272.

We would like to hide the signing key on another machine to onboard more folks to help on infrastructure of NixOS.

# Context

As moving to HSM/PKCS#11 is out of scope in the current timeline we are operating in, we are proposing a simpler solution, albeit very flexible one, which is remote signing technology for Nix.

For this, we need to generalize the signature scheme that assumes we always possess the private key material.

Now, signature schemes use an abstract `Signer` object that can be instantiated anywhere
and passed around in the hierarchy for signing relevant objects.

The old local signing scheme is now implemented as the `LocalSigner`.

The abstract `Signer` always require a local `PublicKey` to make verification cheap.

In addition, we introduce another signer: the `RemoteSigner`, adhering to a "Nix remote signing API",
which is a need for the operations of the NixOS.org infrastructure community, to move the signing key location
in a separate piece of infrastructure with increased scrutiny and security.

The Nix remote signing API has 2 endpoints:

- `/sign` : takes as input a string and reply with its ED25519-SHA512 signature
- `/publickey`: reply the public key material

There is no authentication neither authorization, except for maybe a special string (not yet implemented) that enables to ensure we do not let, by mistake, an open signing server which could sign any path and create a bigger disaster.
Such things are left to firewalling and reverse proxies measures.

Impact evaluation on signature speed, management of timeouts (exponential backoff), error handling are all out of scope for now, as we are still working on this with @cole-h and potentially more folks who may join us.

Perl bindings will not contain a remote signing version as it's hard to abstract properly the `store()->signDetached(...)`, for whatever is worth, a store could not have a signer, and we could not sign things with it.
The only way is to peek and downcast to a store known to have signers. Such a change would be invasive.

# TODO list

- [x] Rework `getPublicKey` idea for remote signer to avoid paying a penalty all the time for nothing
- [x] Validate the remote-signing URL format for easier buildup of API endpoints
- [ ] (Integration?) tests
- [x] Binary cache store implementation?
- [x] Communication key (the "secret special string", which is by no means a serious protection, but _defense in depth_!)
- [x] Fix Perl bindings (for Hydra)

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

RaitoBezarius committed:
libstore: generalize signature scheme

Now, signature schemes uses an abstract `Signer` object that can be instantiated anywhere
and passed around in the hierarchy for signing relevant objects.

The old local signing scheme is now implemented as the `LocalSigner`.

The abstract `Signer` always require a local `PublicKey` to make verification cheap.

In a further commit, we will introduce another signer: the `RemoteSigner`, adhering to a "Nix remote signing API",
which is a need for the operations of the NixOS.org infrastructure community, to move the signing key location
in a separate piece of infrastructure with increased scrutiny and security.

RaitoBezarius committed:
libstore/signature: introduce a remote signer implementation

As the signing key is increasingly an important piece of secret and Nix does not use
any form of PKCS#11 API to enable hardware security based signing, a shortcut we propose for the needs
of the NixOS.org infrastructure community is to employ a "notarized service" or "remote signing service"
which will move the signing key material on another machine.

We call the "Nix remote signing server", a piece of software that offers
an HTTP over UNIX domain socket API implementing the endpoint `/sign`
and `/publickey`:

- `/sign` will take as input a string and reply with its ED25519-SHA512 signature
- `/sign-store-path` will take as an input a store path available to the
  signer's Nix store and reply with its ED25519-SHA512 signature.
- `/publickey` will reply with a public key material that can be stored locally

For now, those APIs have no authorization or authentication, these concerns are left to the infrastructure
operations, e.g. firewalling, reverse proxies, etc.

RaitoBezarius committed:
globals: add `remote-signing-paths` setting

This enables a user to configure remote signing paths for remote signing operations.

RaitoBezarius committed:
libstore/local-store: perform remote signing if available

If remote signing is available, remotely sign realisations and path infos.

RaitoBezarius commented on src/libstore/local-store.cc:
```diff
@@ -1802,22 +1802,36 @@ void LocalStore::signRealisation(Realisation & realisation)
 void LocalStore::signPathInfo(ValidPathInfo & info)
 {
-    // FIXME: keep secret keys in memory.
+    // FIXME: keep secret keys in memory
```
will drop this change

RaitoBezarius commented on src/libstore/signature/local-keys.cc:
```diff
@@ -72,8 +73,7 @@ PublicKey::PublicKey(std::string_view s)
 
-bool verifyDetached(const std::string & data, const std::string & sig,
-    const PublicKeys & publicKeys)
+bool verifyDetached(const std::string & data, const std::string & sig, const PublicKeys& publicKeys)
```
will drop this change

RaitoBezarius commented on src/libstore/signature/local-keys.hh:
```diff
@@ -52,11 +52,7 @@ private:
-/* Return true iff ‘sig’ is a correct signature over ‘data’ using one
-   of the given public keys. */
-bool verifyDetached(const std::string & data, const std::string & sig,
-    const PublicKeys & publicKeys);
```
will drop this change

cole-h commented on src/libstore/globals.hh:
```diff
@@ -638,6 +638,13 @@ public:
           can add it to `trusted-public-keys` in their `nix.conf`.
         )"};
 
+    Setting<Strings> remoteSigningUrls{
```
At some point, we may want to specify (in docs somewhere) how these servers should work (with RFC-style `MUST`s littered everywhere :laughing: ), but of course that can wait until the idea / approach is approved / until _we_ know what we want the server to be able to do.

RaitoBezarius replied:
Addressed in the doc commit, it's a reasonable yet simple attempt I believe for now. We can upgrade this in RFC style later on.

cole-h commented on src/libstore/signature/signer.cc:
```diff
@@ -0,0 +1,85 @@
+
+    std::string RemoteSigner::signDetached(std::string_view s) const {
+        std::string detached_signature;
+        std::string hash_to_sign = std::string(s);
```
Nit:

in other places this is called a fingerprint, so maybe let's keep that nomenclature to prevent confusion (especially since one of the "fingerprints" looks like `1;/nix/store/.........`, and the other is a JSON blob)

lheckemann commented on src/libstore/signature/signer.cc:
```diff
@@ -0,0 +1,85 @@
+    Signer::Signer() : pubkey(this->getPublicKey()) { }
+    Signer::Signer(PublicKey && pubkey) : pubkey(pubkey) { }
+
+    bool Signer::verifyDetached(const std::string & data, const std::string & sig) {
```
Could these be `string_view`s?

lheckemann commented on src/libstore/signature/signer.cc:
```diff
@@ -0,0 +1,85 @@
+            throw Error("Failed to initialize curl");
+        }
+
+        curl_easy_setopt(_curl_handle.get(), CURLOPT_USERAGENT, "Nix $version");
```
We should probably be doing the same as libstore here:
https://github.com/NixOS/nix/blob/78e886bc5fd9e4d85f8503799540c0b71bb270be/src/libstore/filetransfer.cc#L297-L299

lheckemann replied:
(and sharing the same code)

lheckemann replied:
Maybe we should even use FileTransfer here.

RaitoBezarius replied:
I wanted to avoid using FileTransfer because it seems very tailored to downloading/uploading and this is not what we are doing, the thing is quite massive, and I fear that wiring this up would make things quite hard for me.

RaitoBezarius replied:
I proposed a change here, if you want me to use FileTransfer, I would probably need some sync time to discuss how we want to proceed because there would be some stuff to do here.

lheckemann commented on src/libstore/signature/signer.cc:
```diff
@@ -0,0 +1,85 @@
+        curl_easy_setopt(_curl_handle.get(), CURLOPT_TIMEOUT, 5);
+
+        // The remote URL should not have no trailing /.
+        assert(serverUrl.back() != '/');
```
Probably throw an error here, rather than using assert (which can compile to no code!)

lheckemann commented on src/libstore/signature/signer.cc:
```diff
@@ -0,0 +1,85 @@
+        CURLcode res = curl_easy_perform(_curl_handle.get());
+        if (res != CURLE_OK) {
+            // log the error?
+            throw Error("failed to sign remotely, received non-200 code");
```
curl doesn't validate the HTTP response code, this is for stuff like the connection breaking or the server not speaking HTTP.

RaitoBezarius replied:
Fixed.

lheckemann reviewed (commented):
Love to see things happening here!

I wonder if this is fundamentally the right approach, as opposed to sticking to local signatures (with a per-machine key, and the signing done by another machine based on its trust of those per-machine keys). Such an approach would provide somewhat better traceability.

RaitoBezarius commented:
> I wonder if this is fundamentally the right approach, as opposed to sticking to local signatures (with a per-machine key, and the signing done by another machine based on its trust of those per-machine keys). Such an approach would provide somewhat better traceability.

Wouldn't this require copying the paths to the signing box, then getting those paths back, incurring a potential high network traffic?

lheckemann commented:
No, it could take the fingerprint as signed by the builder and re-sign that.

RaitoBezarius commented:
> No, it could take the fingerprint as signed by the builder and re-sign that.

Wouldn't this imply making Nix a signing server itself?

lheckemann commented:
This functionality wouldn't have to be part of Nix. It could just as well be implemented in a cache server like attic, or as a service that uses Nix to copy and re-sign narinfos from an intermediate flat-file binary cache to the "final" one.

RaitoBezarius commented:
> This functionality wouldn't have to be part of Nix. It could just as well be implemented in a cache server like attic, or as a service that uses Nix to copy and re-sign narinfos from an intermediate flat-file binary cache to the "final" one.
> 
> EDIT: and Attic is already 90% of the way there: it stores signatures from the upload, but doesn't currently validate them or deliver them to clients, instead generating a signature using its own per-cache signing key when the narinfo is requested.

Gotcha, but, in practice, for our needs right now, what is the plan you propose with this idea?

I guess this is pushing the problem to an extra piece of software we have to run on the NixOS org machines, right?

Ericson2314 commented:
I think this machinery should go in `libutil` because it is just about signing arbitrary data, and not anything store-specific?

RaitoBezarius commented:
> I think this machinery should go in `libutil` because it is just about signing arbitrary data, and not anything store-specific?

I agree with this.

zimbatm commented:
So move this in libutil and then this can be merged?

Ericson2314 commented:
Yeah if it is moved I support this. Maybe some more rounds of little review after but only little things.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2023-11-17-nix-team-meeting-minutes-104/35753/1

Ericson2314 commented:
Idea sounds good to people in the nix team meeting. Just need to fix the more minor things addressed in the thread.

RaitoBezarius committed:
chore: move signature code from libstore to libutil

As it makes more sense to make it available to more consumers.

In this commit, we create a new util: `randomSha1` because `libstore`
depended on sodium to do that in the past. We moved the sodium
initialization and dependency to `libutil`.

Moreover, we keep `getDefaultPublicKeys` inside `libstore` as it relies
on the global variable `settings` to derive this information and is
really only but a consumer of the `libutil`'s signature code.

RaitoBezarius committed:
libutil/signature: use `string_view` instead of references to string

As we are only reading the string data, this is more idiomatic.

RaitoBezarius committed:
doc: add some words about signatures

We document the format and the remote signature protocol.

RaitoBezarius commented:
I'm converting this stuff to UNIX domain socket-only because after talking with @edef1c we concluded that it was important to be misuse resistant from the right get go. This has no consequence on the caching server because it's very easy to (a) run it directly as a UNIX domain listener (b) use a forwarder properly.

For our production usage, we will use such a mechanism anyway to protect the signing server.

RaitoBezarius committed:
libstore/binary-cache-store: perform remote signing if available

The rest should work out of the box.

RaitoBezarius committed:
libstore/binary-cache-store: add `remote-signing-path` setting

This allows someone oeprating a binary cache store to setup a remote
signing path for remote signing operations.

RaitoBezarius committed:
libstore/signature: memoize the public key in the remote signer

This avoids useless roundtrips every time we want to inquire about the
public key.

RaitoBezarius commented:
The macOS fail is alas unrelated to me :/.

RaitoBezarius commented:
To handle better store signatures, I filed https://github.com/NixOS/hydra/issues/1331 regarding Hydra usage of Perl bindings.

Ericson2314 commented on doc/manual/src/protocols/signature.md:
```diff
@@ -0,0 +1,47 @@
+# Signature
```
`SUMMARY.md` needs to be updated I think?

Ericson2314 commented on src/libutil/hash.hh:
```diff
@@ -143,6 +142,10 @@ public:
     }
 
     static Hash dummy;
```
```suggestion
    static Hash dummy;

```

Ericson2314 commented on src/libutil/signature/local-keys.cc:
```diff
@@ -73,8 +72,7 @@ PublicKey::PublicKey(std::string_view s)
 
-bool verifyDetached(const std::string & data, const std::string & sig,
-    const PublicKeys & publicKeys)
+bool verifyDetached(const std::string & data, const std::string & sig, const PublicKeys& publicKeys)
```
```suggestion
bool verifyDetached(const std::string & data, const std::string & sig, const PublicKeys & publicKeys)
```

Ericson2314 commented on src/libstore/keys.hh:
```diff
@@ -0,0 +1,31 @@
+#pragma once
+///@file
```
```suggestion
///@file

```

Ericson2314 commented on src/libstore/keys.hh:
```diff
@@ -0,0 +1,31 @@
+#include "globals.hh"
+#include "libutil/signature/local-keys.hh"
+
+namespace nix {
```
```suggestion
namespace nix {

```

Ericson2314 commented on src/libutil/signature/signer.cc:
```diff
@@ -0,0 +1,120 @@
+#include <curl/curl.h>
+#include <sodium.h>
+
+namespace nix {
```
We usually do not indent things within namespaces, since it results in basically the whole file being indented.

Ericson2314 commented on src/libutil/util.cc:
```diff
@@ -1,15 +1,23 @@
 
 namespace nix {
 
+const std::string nixVersion = PACKAGE_VERSION;
+
```
Is this used?

RaitoBezarius replied:
Well, yes, the symbol needs to be defined in that library? It's used for `set_user_agent`.

Ericson2314 commented on src/libutil/util.cc:
```diff
@@ -1,15 +1,23 @@
 #include "util.hh"
 #include "fmt.hh"
+#include "sync.hh"
+#include "finally.hh"
+#include "serialise.hh"
+#include "cgroup.hh"
+#include "hash.hh"
```
I think this is unused, from resolving merge conflict

Ericson2314 commented on src/libutil/util.hh:
```diff
@@ -14,8 +14,16 @@
 
 namespace nix {
 
+struct Hash;
+
 void initLibUtil();
 
+/**
+ * This Nix version.
+ */
+extern const std::string nixVersion;
+
+
```
no changes to this file are needed

RaitoBezarius replied:
Why? How do I access `nixVersion` ?

RaitoBezarius committed:
fixup! chore: move signature code from libstore to libutil

Ericson2314 commented on doc/manual/src/protocols/signature.md:
```diff
@@ -0,0 +1,47 @@
+Store path signatures are ED25519 signatures, there are two types of
+signatures:
+
+- `ValidPathInfo` will assemble a fingerprint in the form of `1;<store
+  path>;<NAR hash in base32 including its own type>;<NAR size in bytes>;<NAR
+  references separated by ,>` and sign this information with the ED25519
+  private key.
+- `Realisation` will assemble a fingerprint in the form of a JSON string: `{
+  id, outPath, dependentRealisations }` and sign this information with the
+  ED25519 private key.
+
+# Remote signature protocol
+
+The remote signature protocol is a mechanism to offload the signature of your
+store paths to another machine that can possess the secret key material in a
+secure way.
+
+In this setup, Nix will contact a remote signing URL that you specified and ask
+to sign fingerprints over the wire.
+
+The protocol expects a UNIX domain socket to force you to handle proper
+authentication and authorization. `socat` is a great tool to manipulate all
+sorts of sockets.
+
+## Semantics of the APIs
+
+- `POST /sign`: expects a fingerprint as input and will return the signature
+  associated to that fingerprint.
+- `POST /sign-store-path`: expects a store path as a parameter and will attempt
+  to sign that specific store path which is expected to be present on the
+  signer's machine and return the signature in the response.
+- `GET /publickey`: receives the public key as a string in the response.
+
+## A note on the security of that scheme
+
+You are responsible to ensure that `/sign` cannot be abused to sign anything
+and everything, for this, a simple setup could involve setting up a TCP service
+that requires authentication, e.g. SSH or something on the HTTP level and you
+can run a highly privileged daemon on the machine that wants to benefit from
+signatures presenting a UNIX domain socket to Nix.
```
This file should be one sentence per line (or other punctuation / semantic boundary), not fixed line length wrapped.

RaitoBezarius replied:
Does that mean that you want to remove the fixed line length wrapping and leave it the existing punctuation or ?

Ericson2314 replied:
I'll do it. @fricklerhandwerk wrote the non-fixed length lines somewhere in the contributions docs but I forget where. You can find it if you are curious.

Ericson2314 committed:
Merge remote-tracking branch 'upstream/master' into signature-generalization

Ericson2314 committed:
Merge remote-tracking branch 'upstream/master' into signature-generalization

Master included #9688 prep PR, shrinking the diff. Also the merge
includes a number of slight adjustments from @Ericson2314.

Ericson2314 commented on doc/manual/src/protocols/signature.md:
```diff
@@ -0,0 +1,49 @@
+
+  Expects a fingerprint as input and will return the signature associated to that fingerprint.
+
+- `POST /sign-store-path`:
+
+  Expects a store path as a parameter and will attempt to sign that specific store path which is expected to be present on the signer's machine and return the signature in the response.
```
This doesn't exist at the moment, but I wonder whether it should. Yes, I talked about moving things into `libutil` before, but my experience with hardware parse-display-and-sign devices make me wonder if we should always send over a "nice, JSON" form of what is to be signed, and the other end is responsible for concatenating the actual signature data.

If nothing else, forcing the remote side to munge the bytes is a good defense against lazy blind-singing implementations.

cole-h replied:
Well, it "exists" in that I implemented it in my service-side of this PR (https://github.com/cole-h/nixos-cache-signing-server), heh.

I do think sending the bits necessary for signing as a well-formed JSON document is a good idea, and will make it harder to misuse.

(The only reason `/sign-store-path` exists is because I initially implemented it in terms of shelling out to `nix`, because at the time I couldn't figure out how to do the signing as Nix does it. This has been since rectified, so honestly I wouldn't be opposed to removing it altogether, unless there's a reason to keep it around?)

RaitoBezarius replied:
As said privately to @Ericson2314, I don't believe this sign store path API is good for us yet because we are looking at a high frequency signature system, e.g. hydra.nixos.org and there's no distributed filesystems that will make those stores paths available to the signer in time so they can sign them.

For technical reasons, we need a blind signer that is sufficiently resistant to misuse to avoid disasters for random people and ourselves.

Sending a well-formed JSON document looks good to me, but how is it harder to misuse? The only way to make it better is to introduce authnz or auth inside the server and the protocol, this was obviously out of scope because we cannot represent all the usecases in such a simple PR.

Once this lands, my opinion is that we should focus our efforts on understanding how to move the signing to the remote server using PKCS#11 API (which is almost ready with `cryptoki` and `rust-cryptoki` thanks to @baloo) and then think from the HSM perspective. There are many missing pieces and knobs around this system of signature: no revocation, no "timestamping", no certification paths, etc. This is understandable why but all those years have passed and those missing features are limiting our operational range in regards to the cache security.

fricklerhandwerk commented on doc/manual/src/protocols/signature.md:
```diff
@@ -0,0 +1,49 @@
+
+In this setup, Nix will contact a remote signing URL that you specified and ask to sign fingerprints over the wire.
+
+The protocol expects a UNIX domain socket to force you to handle proper authentication and authorization.
+`socat` is a great tool to manipulate all sorts of sockets.
```
This is missing an example of how any of this looks in action. Right now I wouldn't even know where to start in order to use it.

RaitoBezarius replied:
I wonder if I should really give an example because I don't want to give the trivial example that will make people run it insecurely, so I can give non-trivial examples to run it securely but with imaginary TCP services.

Ericson2314 replied:
Yeah I plan on revising the docs a good bit

fricklerhandwerk replied:
It would be enough to show how to invoke Nix to listen on a socket. I don't see from the docs which interfaces this change even exposes.

mschwaig commented:
> I wonder if this is fundamentally the right approach, as opposed to sticking to local signatures (with a per-machine key, and the signing done by another machine based on its trust of those per-machine keys). Such an approach would provide somewhat better traceability.

I agree and would like to add to this point. I see it as a big design issue in the existing signing scheme that the signatures do not make any statements about the origin of a path.
An updated signature format could include a cryptographically verifiable statement that the signing host (claims to have) built a given path itself. I think such a signature would serve as a good basis for another host (remote signer) to also produce a signature for the given path.

See https://github.com/NixOS/nix/issues/9644 for a more detailed description of that idea.

baloo commented on doc/manual/src/protocols/signature.md:
````diff
@@ -0,0 +1,49 @@
+
+- `ValidPathInfo` will assemble a fingerprint in the form of
+  ```
+  1;<store path>;<NAR hash in base32 including its own type>;<NAR size in bytes>;<NAR references separated by ,>
+  ```
+  and sign this information with the ED25519 private key.
````
I'm not sure what the exact signature scheme is, but if we intend to sign prehashed payload (and use the `NAR hash in base32` specified here), ed25519 requires it to be sha512.
(https://docs.rs/ed25519-dalek/latest/ed25519_dalek/trait.DigestSigner.html#impl-DigestSigner%3CD,+Signature%3E-for-Context%3C'_,+'_,+SigningKey%3E)

tomberek commented:
stale: re-open and re-base as needed
